src/tpm2_createpolicy.c

```diff
@@ -0,0 +1,330 @@
+ *      Author: idesai
+ */
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <sys/stat.h>
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+#include <ctype.h>
+#include <getopt.h>
+
+#include <sapi/tpm20.h>
+#include <tcti/tcti_socket.h>
+
+#include "common.h"
+#include "sample.h"
+#include "log.h"
+#include "options.h"
+#include "main.h"
+#include "string-bytes.h"
+#include "tpm_session.h"
+#include "tpm_hash.h"
+
+
+typedef struct createpolicypcr_ctx createpolicypcr_ctx;
+struct createpolicypcr_ctx {
+    TSS2_SYS_CONTEXT *sapi_context;
+    SESSION *policy_session;
+    TPM2B_DIGEST policy_digest;
+    char policyfile[PATH_MAX];
+    //build_pcr_policy
+    unsigned int pcr_index;
+    TPMI_ALG_HASH hash_alg;
+    char raw_pcr_file[PATH_MAX];
+    unsigned raw_pcr_flag;
+};
+
+#define INIT_SIMPLE_TPM2B_SIZE( type ) (type).t.size = sizeof( type ) - 2;
+
+#define SET_PCR_SELECT_BIT( pcrSelection, pcr ) \
+ (pcrSelection).pcrSelect[( (pcr)/8 )] |= ( 1 << ( (pcr) % 8) );
+
+TPM_RC build_pcr_policy( createpolicypcr_ctx *ctx) {
+
+    TPM_RC rval = TPM_RC_SUCCESS;
+    TPML_PCR_SELECTION pcrs;
+    TPML_DIGEST pcrValues;
+    UINT32 pcrUpdateCounter;
+    TPM2B_DIGEST pcrDigest;
+    TPML_PCR_SELECTION pcrSelectionOut;
+    pcrs.count = 1;
+    pcrs.pcrSelections[0].hash = ctx->hash_alg;
+    pcrs.pcrSelections[0].sizeofSelect = 3; 
+    pcrs.pcrSelections[0].pcrSelect[0] = 0; 
+    pcrs.pcrSelections[0].pcrSelect[1] = 0; 
+    pcrs.pcrSelections[0].pcrSelect[2] = 0; 
+    SET_PCR_SELECT_BIT( pcrs.pcrSelections[0], ctx->pcr_index); 
+
+    FILE *fp;
+    struct stat st;
+    int errno =0, i=0, sz=0;
+    if(ctx->raw_pcr_flag) {
+        //digests[count{2:}]{:8}
+        pcrValues.count = pcrs.count;
+
+        switch(ctx->hash_alg) {
+            case TPM_ALG_SHA1: 
+                pcrValues.digests[0].t.size = SHA1_DIGEST_SIZE;
+                break;
+            case TPM_ALG_SHA256:
+                pcrValues.digests[0].t.size = SHA256_DIGEST_SIZE;
+                break;
+            case TPM_ALG_SHA384:
+                pcrValues.digests[0].t.size = SHA384_DIGEST_SIZE;
+                break;            
+            case TPM_ALG_SHA512:
+                pcrValues.digests[0].t.size = SHA512_DIGEST_SIZE;
+                break;
+            case TPM_ALG_SM3_256:
+                pcrValues.digests[0].t.size = SM3_256_DIGEST_SIZE;
+                break;
+            default:
+                pcrValues.digests[0].t.size = SHA1_DIGEST_SIZE;
+                break;
+        }
+        if (stat(ctx->raw_pcr_file, &st) == 0) {
+            if(st.st_size != pcrValues.digests[0].t.size) {
+                printf("pcr input file has wrong size.\n");
+                return -1;
+            }
+        }
+        else {
+            fprintf(stderr, "Cannot determine size of %s: %s\n",
+            ctx->raw_pcr_file, strerror(errno));
+            return -1;
+        }
+        fp = fopen(ctx->raw_pcr_file, "rb");
```

Comment:
Actually since you're opening the file, you could use the getFileSize() routine in files.h and avoid using stat which is not portable.